### PR TITLE
Adding data type information for data-driven intents.

### PR DESCRIPTION
**Pull request by nkolba**
This PR is to address the handling of data-driven intents that can be raised in current API draft spec.  By providing a place to specify type for both payload of input and output of an  intent, we give the scaffolding for desktop agents to perform the following functions:

- Manage the timing of resolving the Promise for a raised intent.  If an intent has no data payload, it should be resolved as soon as it is routed.  If an intent has a data return defined, the desktop agent should wait to resolve until the handling app has returned the (valid) data.
- Validate data payloads, both when an intent is raised and is being routed (if there is a type specified for the input), and when resolving.

### CODE REVIEW

File: src/Intent.yaml [hunk 1]
```diff
@@ -23,6 +23,16 @@ Intent:
       description: The contexts the intent accepts. This will
         typically be a set of namespaced context types, e.g. 
         "org.symphony.contact"
+    dataReturnType: 
+      type: string
+      description: If there is a data payload associated with the intent resolution, specifies the type.  
+      This value can be either a URL to the type definition, or an alias for the type.  
+      If this value is set, the desktop agent can manage the resolution of the intent accordingly.
```
Comment (rikoe) [inline src/Intent.yaml:30]: I am not a fan of this. The difference between a data-driven intent and fire-and-forget intent is then implicit based on whether this property is present or not. Lots of opportunity for misunderstanding.

I would like it to be more explicit: An "intentType" of some kind that indicates the difference, or at the top level a different object for "intent".

File: src/Intent.yaml [hunk 2]
```diff
@@ -23,6 +23,16 @@ Intent:
...
+    dataInputType: 
+      type: string
+      description: If there is a data payload associated with the intent, specifies the type expected.  
+      This value can be either a URL to the type definition, or an alias for the type.  
+      If this value is set, the desktop agent can manage the validation and resolution of the intent accordingly.
```
Comment (rikoe) [inline src/Intent.yaml:35]: This is a major divergence from the spec so far. We currently use the "contexts" array for the purpose of specifying the possible "input types" (plural) that an intent can accept/operate on. How do these two interact?

File: Specification-Draft.MD [hunk 3]
```diff
@@ -20,7 +20,7 @@ FDC3 Intents define a standard set of nouns and verbs that can be used to put to
 * Atomic
     * There should be clear lines in a workflow where the object begins and ends
 * Stateless
-    * Workflows should not require callbacks or endpoints to maintain references to each other.  Once an intent is passed to an endpoint - it controls the rest of that workflow. 
+    * Workflows should not require callbacks or endpoints to maintain references to each other.  With the exception of specifically data-driven intents, once an intent is passed to an endpoint - it controls the rest of that workflow.
```
Comment (rikoe) [inline Specification-Draft.MD:23]: Just saying this isn't enough, in my opinion. The doc needs to clearly set out the difference between the two.

**Review by rikoe — CHANGES_REQUESTED** [review]

**Review by espenove — COMMENTED** [review]
I rather we follow the approach used by Android Intents, where the initiating application uses a separate API when a result is wanted:

No result: startActivity()
Wait for result: startActivityForResult()  

https://developer.android.com/guide/components/intents-filters

**Comment by espenove** [comment]
Discussed on Intents WG meeting - https://finosfoundation.atlassian.net/wiki/spaces/FDC3/pages/553746496/2018-10-03+Meeting+agenda
Agreed to not move forward with proposed change

**Pull request closed without merging**